core/src/test/java/org/testcontainers/utility/AuthenticatedImagePullTest.java

```diff
@@ -0,0 +1,117 @@
+package org.testcontainers.utility;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.model.AuthConfig;
+import com.github.dockerjava.core.command.PullImageResultCallback;
+import com.github.dockerjava.core.command.PushImageResultCallback;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.testcontainers.DockerClientFactory;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.HttpWaitStrategy;
+import org.testcontainers.images.builder.ImageFromDockerfile;
+
+import java.util.concurrent.TimeUnit;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.when;
+import static org.rnorth.visibleassertions.VisibleAssertions.assertTrue;
+
+/**
+ * This test checks the integration between Testcontainers and an authenticated registry, but uses
+ * a mock instance of {@link RegistryAuthLocator} - the purpose of the test is solely to ensure that
+ * the auth locator is utilised, and that the credentials it provides flow through to the registry.
+ *
+ * {@link RegistryAuthLocatorTest} covers actual credential scenarios at a lower level, which are
+ * impractical to test end-to-end.
+ */
+public class AuthenticatedImagePullTest {
+
+    @ClassRule
+    public static GenericContainer authenticatedRegistry = new GenericContainer(new ImageFromDockerfile()
+        .withDockerfileFromBuilder(builder -> {
+            builder.from("registry:2")
+                .run("htpasswd -Bbn testuser notasecret > /htpasswd")
+                .env("REGISTRY_AUTH", "htpasswd")
+                .env("REGISTRY_AUTH_HTPASSWD_PATH", "/htpasswd")
+                .env("REGISTRY_AUTH_HTPASSWD_REALM", "Test");
+        }))
+        .withExposedPorts(5000)
+        .waitingFor(new HttpWaitStrategy());
+
+    private static RegistryAuthLocator originalAuthLocatorSingleton;
+    private static DockerClient client;
+
+    private static String testRegistryAddress;
+    private static String testImageName;
+    private static String testImageNameWithTag;
+
+    @BeforeClass
+    public static void setUp() {
+        originalAuthLocatorSingleton = RegistryAuthLocator.instance();
+        client = DockerClientFactory.instance().client();
+
+        testRegistryAddress = authenticatedRegistry.getContainerIpAddress() + ":" + authenticatedRegistry.getFirstMappedPort();
+        testImageName = testRegistryAddress + "/alpine";
+        testImageNameWithTag = testImageName + ":latest";
```

Comment:
not important: If we use random image name here, we will not have to remove it in `tearDown`

Reply:
I think we'd still have to remove it wouldn't we? I've left as anyway, since what is here is still safe.